src/libstore/build/worker.cc

```diff
@@ -123,24 +154,46 @@ GoalPtr Worker::makeGoal(const DerivedPath & req, BuildMode buildMode)
 }
 
 
+template<typename K, typename V, typename F>
+static void filterMap(std::map<K, V> & goalMap, F f)
```

Comment:
```suggestion
static void cullMap(std::map<K, V> & goalMap, F f)
```

It's not an immutable filter but a destructive ~mutilation~ mutation.